_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,699 @@
+ARMv6-M architecture
+
+## Advanced Exception Topics
+
+### Exception Entry & Exit
+
+One of my favorite parts about ARM exception entry is that the hardware itself implements the ARM
+Architecture Procedure Calling Standard (AAPCS) [^5] which defines a set of conventions that must
+be followed by compilers. One of these requirements is around the registers which must be saved by
+a C function when calling another function. When an exception is invoked, the hardware will
+automatically stack the registers that are caller-saved. The hardware will then encode in the link register (`$lr`) a value known as the `EXC_RETURN` value. This value tells the ARM core that a return from an exception is taking place and the core can then unwind the stack and return correctly to the thread mode code which was running before the exception took place
+
+By leveraging these features, exceptions and thread mode code can share the same set of registers and exception entries can be regular C functions! For other architectures, exception handlers often have to be written in assembly.
+
+### Tail-Chaining
+
+As discussed above, usually when exiting an exception, the hardware needs to pop and
+restor at least eight caller-saved registers. However, when exiting an ISR while a new ISR is already
+pended, this pop and subsequent push of the same register state can be skipped since it would be
+popping and then pushing the exact same registers! This optimization
+is knoown as "Tail-Chaining".
+
+For example, on a Cortex-M3, when using zero
+wait state memory, it takes 12 clock cycles to start executing an ISR after it has been asserted
+and 12 cycles return from the ISR upon its completion. When the register pop and push is skipped,
+it only takes 6 cycles to exit from one exception and start another one, saving 18 cycles in total!
+
+### Late-arriving Preemption
+
+The ARM core can detect a higher priority exception while in the "exception entry phase"
+(stacking caller registers & fetching the ISR routine to execute). A "late arriving" interrupt is detected during
+this period. The optimization is that the higher priority ISR can be fetched and executed but the
+register state saving that has already taken place can be skipped. This reduces the latency for the
+higher priority interrupt and, conveniently, upon completion of
+the late arriving exception handler, the processor can then tail-chain into the initial exception
+that was going to be serviced.
+
+### Lazy State Preservation
+
+ARMv7 & ARMv8 devices can implement an optional **Floating Point Unit** (`FPU`) for native floating
+point support. This comes with the addition of 33 4 byte registers (`s0`-`s31` & `fpscr`). 17 of
+these are "caller" saved and need to be dealt with by the ARM exception entry handler. Since FPU
+registers are not often used in ISRs, there is an optimization that can be enabled which defers the
```

Comment:
It would be nice to add a brief sentence about *how* that optimization is enabled.